chain/src/txhashset/bitmap_accumulator.rs

```diff
@@ -0,0 +1,241 @@
+		}
+
+		// Trim any "empty" chunks from the right to keep everything consistent.
+		self.trim_right()?;
+
+		Ok(())
+	}
+
+	/// Given the provided chunk idx rewind the bitmap accumulator to the end of the
+	/// previous chunk ready for the updated chunk to be appended.
+	fn rewind_prior(&mut self, chunk_idx: u64) -> Result<(), Error> {
+		let last_pos = self.backend.size();
+		let mut pmmr = PMMR::at(&mut self.backend, last_pos);
+		let chunk_pos = pmmr::insertion_to_pmmr_index(chunk_idx + 1);
+		let rewind_pos = chunk_pos.saturating_sub(1);
+		pmmr.rewind(rewind_pos, &Bitmap::create())
+			.map_err(|e| ErrorKind::Other(e))?;
+		Ok(())
+	}
+
+	/// Make sure we append empty chunks to fill in any gap before we append the chunk
+	/// we actually care about. This effectively pads the bitmap with 1024 chunks of 0s
+	/// as necessary to put the new chunk at the correct place.
+	fn pad_left(&mut self, chunk_idx: u64) -> Result<(), Error> {
+		let last_pos = self.backend.size();
+		let chunk_pos = pmmr::insertion_to_pmmr_index(chunk_idx);
+		for _ in last_pos..chunk_pos {
+			self.append_chunk(BitmapChunk::new())?;
+		}
+		Ok(())
+	}
+
+	/// Recursively trim the last chunk if empty. Stop when rightmost chunk is non-empty
+	/// or the accumulator itself is empty.
+	fn trim_right(&mut self) -> Result<(), Error> {
+		let last_pos = self.backend.size();
+		if last_pos == 0 {
+			return Ok(());
+		}
+		let mut pmmr = PMMR::at(&mut self.backend, last_pos);
+		if pmmr.get_hash(last_pos)
+			== Some(BitmapChunk::new().hash_with_index(last_pos.saturating_sub(1)))
+		{
+			pmmr.rewind(last_pos.saturating_sub(1), &Bitmap::create())
+				.map_err(|e| ErrorKind::Other(e))?;
+			self.trim_right()
+		} else {
+			Ok(())
+		}
+	}
+
+	/// Append a new chunk to the BitmapAccumulator.
+	/// Append parent hashes (if any) as necessary to build associated peak.
+	pub fn append_chunk(&mut self, chunk: BitmapChunk) -> Result<u64, Error> {
+		let last_pos = self.backend.size();
+		PMMR::at(&mut self.backend, last_pos)
+			.push(&chunk)
+			.map_err(|e| ErrorKind::Other(e).into())
+	}
+
+	/// The root hash of the bitmap accumulator MMR.
+	pub fn root(&self) -> Hash {
+		ReadonlyPMMR::at(&self.backend, self.backend.size()).root()
+	}
+}
+
+/// A bitmap "chunk" representing 1024 contiguous bits of the overall bitmap.
+/// The first 1024 bits belong in one chunk. The next 1024 bits in the next chunk, etc.
+#[derive(Clone, Debug)]
+pub struct BitmapChunk(BitVec);
+
+impl BitmapChunk {
+	const LEN_BITS: usize = 1024;
+	const LEN_BYTES: usize = Self::LEN_BITS / 8;
+
+	/// Create a new bitmap chunk, defaulting all bits in the chunk to false.
+	pub fn new() -> BitmapChunk {
+		BitmapChunk(BitVec::from_elem(Self::LEN_BITS, false))
+	}
+
+	/// Set a single bit in this chunk.
+	/// 0-indexed from start of chunk.
+	pub fn set(&mut self, idx: u64, value: bool) {
+		let idx = usize::try_from(idx).expect("usize from u64");
+		if idx < Self::LEN_BITS {
```

Comment:
what if idx >= LEN_BITS? if that should never happen, then we should assert that instead.
silently ignoring it looks bad.

Reply:
Yes agreed. The silent ignore is bad practice.